### PR TITLE
fix: serialize eth_subscribe log data

### PR DESCRIPTION
**Pull request by MicaiahReid**

### CODE REVIEW

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 1]
```diff
@@ -244,4 +245,40 @@ describe("provider", () => {
...
+    let web3: Web3;
+    let accounts: string[];
+
+    beforeEach(async () => {
+      provider = await getProvider({
+        chain: { networkId },
+        wallet: { deterministic: true }
+      });
+      web3 = new Web3();
+      // TODO: remove "as any" once we've fixed our typing issues with web3
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: Can you add a reference to the PRs that will fix this?

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 2]
```diff
@@ -244,4 +245,40 @@ describe("provider", () => {
...
+      accounts = await web3.eth.getAccounts();
+    });
+
+    it("returns things via legacy", async () => {
+      const subscription = web3.eth.subscribe("newBlockHeaders");
+      assert(subscription != null);
+      assert.notStrictEqual(subscription.id, false);
+
+      // if the data isn't properly serialized before emitting, web3 won't
+      // ever catch this
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: What do you mean by "catch" here? Would "handle" be a better word?

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 3]
```diff
@@ -244,4 +245,40 @@ describe("provider", () => {
...
+
+  describe("web3 compatibility", () => {
+    const networkId = 1234;
+    let provider: EthereumProvider;
+    let web3: Web3;
+    let accounts: string[];
+
+    beforeEach(async () => {
+      provider = await getProvider({
+        chain: { networkId },
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: Do we need to set a `networkId` here? If not I think it should be removed.

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 4]
```diff
@@ -244,4 +245,40 @@ describe("provider", () => {
...
+
+    it("returns things via legacy", async () => {
+      const subscription = web3.eth.subscribe("newBlockHeaders");
+      assert(subscription != null);
+      assert.notStrictEqual(subscription.id, false);
+
+      // if the data isn't properly serialized before emitting, web3 won't
+      // ever catch this
+      subscription.on("data", data => {
+        assert.strictEqual(
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: I think this test will still pass even if `data` handler isn't handled.

File: src/chains/ethereum/ethereum/src/provider.ts [hunk 5]
```diff
@@ -359,13 +359,14 @@ export default class EthereumProvider
     >;
     if (promise instanceof PromiEvent) {
       promise.on("message", data => {
+        const serialized = JSON.parse(JSON.stringify(data));
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/provider.ts]: I don't think `serialized` is the right name here. Maybe `normalizedData` or `clonedData` would be better?

File: src/chains/ethereum/ethereum/src/api.ts [hunk 6]
```diff
@@ -2054,13 +2054,10 @@ export default class EthereumApi implements Api {
...
             (result as any).baseFeePerGas = header.baseFeePerGas;
           }
-
-          // TODO: move the JSON stringification closer to where the message
-          // is actually sent to the listener
           promiEvent.emit("message", {
             type: "eth_subscription",
             data: {
-              result: JSON.parse(JSON.stringify(result)),
+              result: result,
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/api.ts]:
```suggestion
              result,
```

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 7]
```diff
@@ -244,4 +245,40 @@ describe("provider", () => {
...
+        wallet: { deterministic: true }
+      });
+      web3 = new Web3();
+      // TODO: remove "as any" once we've fixed our typing issues with web3
+      web3.setProvider(provider as any);
+      accounts = await web3.eth.getAccounts();
+    });
+
+    it("returns things via legacy", async () => {
+      const subscription = web3.eth.subscribe("newBlockHeaders");
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: How does this work without awaiting it first? :thinking:

**Review by davidmurdoch — CHANGES_REQUESTED** [review]

File: src/chains/ethereum/ethereum/src/provider.ts (hunk 5 above)
Comment (MicaiahReid) [reply src/chains/ethereum/ethereum/src/provider.ts]: I like `normalizedData`. Updated!

File: src/chains/ethereum/ethereum/tests/provider.test.ts (hunk 3 above)
Comment (MicaiahReid) [reply src/chains/ethereum/ethereum/tests/provider.test.ts]: Removed.

File: src/chains/ethereum/ethereum/tests/provider.test.ts (hunk 1 above)
Comment (MicaiahReid) [reply src/chains/ethereum/ethereum/tests/provider.test.ts]:
```suggestion
      // TODO: remove "as any" once we've fixed our typing issues
      // with web3 (https://github.com/ChainSafe/web3.js/pull/4761)
```

File: src/chains/ethereum/ethereum/tests/provider.test.ts (hunk 2 above)
Comment (MicaiahReid) [reply src/chains/ethereum/ethereum/tests/provider.test.ts]: Yes, I definitely wasn't thinking of "catch" as an actual technical term here. Handle is better.
```suggestion
      // ever handle this
```

File: src/chains/ethereum/ethereum/tests/provider.test.ts (hunk 4 above)
Comment (MicaiahReid) [reply src/chains/ethereum/ethereum/tests/provider.test.ts]: Fixed test!

**Review by davidmurdoch — APPROVED** [review]

**Pull request merged**